src/main/kotlin/com/terraformation/backend/admin/AdminCohortsController.kt

```diff
@@ -0,0 +1,169 @@
+import org.springframework.web.bind.annotation.GetMapping
+import org.springframework.web.bind.annotation.PathVariable
+import org.springframework.web.bind.annotation.PostMapping
+import org.springframework.web.bind.annotation.RequestMapping
+import org.springframework.web.bind.annotation.RequestParam
+import org.springframework.web.servlet.mvc.support.RedirectAttributes
+
+@Controller
+@RequestMapping("/admin")
+@RequireGlobalRole(
+    [GlobalRole.SuperAdmin, GlobalRole.AcceleratorAdmin, GlobalRole.TFExpert, GlobalRole.ReadOnly])
+@Validated
+class AdminCohortsController(
+    private val cohortStore: CohortStore,
+    private val moduleStore: ModuleStore,
+) {
+  private val log = perClassLogger()
+
+  @GetMapping("/cohorts")
+  fun cohortsHome(model: Model): String {
+    val cohorts = cohortStore.findAll()
+
+    val canReadCohorts = cohorts.all { currentUser().canReadCohort(it.id) }
+
+    model.addAttribute("cohorts", cohorts)
+    model.addAttribute("canReadCohorts", canReadCohorts)
+
+    return "/admin/cohorts"
+  }
+
+  @GetMapping("/cohorts/{cohortId}")
+  fun cohortView(
+      model: Model,
+      @PathVariable cohortId: CohortId,
+      redirectAttributes: RedirectAttributes
+  ): String {
+    requirePermissions { readCohort(cohortId) }
+    val cohort =
+        try {
+          cohortStore.fetchOneById(cohortId, CohortDepth.Cohort, CohortModuleDepth.Module)
+        } catch (e: CohortNotFoundException) {
+          log.warn("Cohort not found", e)
+          redirectAttributes.failureMessage = "Cohort not found: ${e.message}"
+          return redirectToCohortHome()
+        }
+    val cohortModules = cohort.modules.map { it.moduleId }
+
+    val modules = moduleStore.fetchAllModules()
+    val moduleNames = modules.associate { it.id to "(${it.id}) ${it.name}" }
+
+    model.addAttribute("cohort", cohort)
+    model.addAttribute("cohortModules", cohortModules)
+    model.addAttribute("moduleNames", moduleNames)
+    model.addAttribute("canUpdateCohort", currentUser().canUpdateCohort(cohortId))
+
+    return "/admin/cohortView"
+  }
+
+  @PostMapping("/cohorts/{cohortId}/addModule")
+  fun addModule(
+      model: Model,
+      @PathVariable cohortId: CohortId,
+      @RequestParam moduleId: ModuleId,
+      @RequestParam startDate: LocalDate,
+      @RequestParam endDate: LocalDate,
+      redirectAttributes: RedirectAttributes
+  ): String {
+    requirePermissions { updateCohort(cohortId) }
+    try {
+      cohortStore.update(cohortId) {
+        val updatedModules =
+            it.modules.plus(CohortModuleModel(cohortId, moduleId, startDate, endDate))
+        it.copy(modules = updatedModules)
+      }
+      redirectAttributes.successMessage = "Cohort module added."
+    } catch (e: Exception) {
+      log.warn("Add cohort module failed")
+      redirectAttributes.failureMessage =
+          when (e) {
+            is DataIntegrityViolationException -> "Add module failed. Dates are invalid."
```

Comment:
Worth explaining the constraints in this error message so people will have a better idea what they did wrong? This will usually be user error rather than a system failure.

When we add this functionality to the real web app, we'll probably want to make the failure checking a little more fine-grained, e.g., by catching this in `CohortStore` and querying to see which modules' dates overlap so we can give the user more specific feedback.

Reply:
I will follow up in a separate PR to add this in `CohortStore`. 

I think since modules may be allowed to overlap, we may remove the dates overlapping constraints. Will updates documentations once all of those requirements are finalized.